### PR TITLE
Added a rudimentary settings option for changing balls. Also implemen…

### LINKED ISSUE #32: Figure out where Ball shape should be set
Should it be set in Settings for single player? Should this preference be carried over to multiplayer, or should multiplayer handle this in an isolated fashion?

### PR DESCRIPTION
**Pull request by Jonas-C**
…ted Soccerballs and Basketballs.

### CODE REVIEW

File: core/src/com/gaalf/game/precreatedEntities/balls/BallFactory.java [hunk 1]
```diff
@@ -23,10 +22,14 @@ public BallFactory(World world, GameAssetManager assetManager){
     @Override
     public Entity createEntity(String type, String playerName, int playerNumber, TiledMap tiledMap) {
         switch (type) {
-            case "golfBall":
+            case "Golfball":
```
Comment (Trulsmatias) [inline core/src/com/gaalf/game/precreatedEntities/balls/BallFactory.java]: why change? I think all the string values was better as they were

File: core/src/com/gaalf/game/precreatedEntities/balls/Ball.java [hunk 2]
```diff
@@ -47,9 +46,9 @@ private SoundComponent addSoundComponent(Sound sound){
         return soundComponent;
     }
 
-    TextureComponent addTextureComponent(Texture texture){
+    TextureComponent addTextureComponent(Sprite sprite){
```
Comment (Trulsmatias) [inline core/src/com/gaalf/game/precreatedEntities/balls/Ball.java]: It is kind of weird that it doesn't take in a texture object any longer considering the name of the function. But I see why you did it

File: core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        Sprite sprite = textureAtlas.createSprite("Basketball");
+        TextureComponent textureComponent = addTextureComponent(sprite);
+
+        CircleShape cshape = new CircleShape();
+        cshape.setRadius(((textureComponent.sprite.getRegionWidth() * transformComponent.scale.x) / 2) / PPM);
+        FixtureDef fixtureDef = new FixtureDef();
+        fixtureDef.shape = cshape;
+        fixtureDef.density = 8;
+        fixtureDef.friction = 100.6f;
+        fixtureDef.restitution = .5f;
```
Comment (Trulsmatias) [inline core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java:30]: Maybe add those values to a file as well, as was done on the other values of the ball?

File: core/src/com/gaalf/game/precreatedEntities/balls/GolfBall.java [hunk 4]
```diff
@@ -16,7 +17,9 @@
     GolfBall(String playerName, int playerNumber, TiledMap tiledMap, World world, GameAssetManager assetManager) {
         super(playerName, playerNumber, assetManager);
         TransformComponent transformComponent = addTransformComponent(tiledMap, 1.5f);
-        TextureComponent textureComponent = addTextureComponent((Texture)assetManager.manager.get(assetManager.golfBallSprite));
+        TextureAtlas textureAtlas = assetManager.manager.get(assetManager.ballSpriteAtlas);
+        Sprite sprite = textureAtlas.createSprite("Golfball");
```
Comment (Trulsmatias) [inline core/src/com/gaalf/game/precreatedEntities/balls/GolfBall.java]: again, doesn't like the naming convention of the balls

File: core/src/com/gaalf/game/precreatedEntities/balls/SquareBall.java [hunk 5]
```diff
@@ -14,7 +15,9 @@
     SquareBall(String playerName, int playerNumber, TiledMap tiledMap, World world, GameAssetManager assetManager) {
         super(playerName, playerNumber, assetManager);
         TransformComponent transformComponent = addTransformComponent(tiledMap, 1.5f);
-        TextureComponent textureComponent = addTextureComponent((Texture)assetManager.manager.get(assetManager.golfBallSprite));
+        TextureAtlas textureAtlas = assetManager.manager.get(assetManager.ballSpriteAtlas);
+        Sprite sprite = textureAtlas.createSprite("Square");
+        TextureComponent textureComponent = addTextureComponent(sprite);
```
Comment (Trulsmatias) [inline core/src/com/gaalf/game/precreatedEntities/balls/SquareBall.java]: Maybe move the creation of the sprite inside the addTextureComponent() function, and just take the string as a parameter to the function instead?

File: core/src/com/gaalf/manager/SettingsManager.java [hunk 6]
```diff
@@ -9,24 +9,38 @@
...
         soundIsEnabled = isSoundEffectsEnabled();
+        ballChoice = getBallChoice();
     }
 
     private Preferences getPreferences(){
         return Gdx.app.getPreferences(PREFS_NAME);
     }
 
+    public String getBallChoice(){
+        return getPreferences().getString(PREF_BALL, "Golfball");
```
Comment (Trulsmatias) [inline core/src/com/gaalf/manager/SettingsManager.java:34]: Why is "Golfball" hardcoded here?

File: core/src/com/gaalf/presenter/MainMenuPresenter.java [hunk 7]
```diff
@@ -17,15 +17,14 @@ public MainMenuPresenter(final GaalfGame game){
...
+        game.playersManager.addPlayer("Jonas", true, game.settingsManager.getBallChoice());
         game.devicePlayer = game.playersManager.getPlayers().get(0);
         game.setScreen(new GamePresenter(game, game.levelManager.getRandomLevel()));
     }
 
     public void openLevelSelectMenu() {
-        game.playersManager.addPlayer("Jonas", true);
-        game.playersManager.addPlayer("E", false);
+        game.playersManager.addPlayer("Jonas", true, game.settingsManager.getBallChoice());
+        game.playersManager.addPlayer("E", false, "Square");
```
Comment (Trulsmatias) [inline core/src/com/gaalf/presenter/MainMenuPresenter.java:27]: looks confusing. Maybe change the string to "squareBall"?

File: core/src/com/gaalf/manager/GameAssetManager.java [hunk 8]
```diff
@@ -16,9 +17,7 @@
     public final String jumpSound = "sounds/JumpSound.wav";
     public final String finishSound = "sounds/FinishSound.wav";
 
-    public final String basketBallSprite = "balls/basketball.png";
-    public final String golfBallSprite = "balls/golfball.png";
-    public final String soccerBallSprite = "balls/soccerball.png";
+    public final String ballSpriteAtlas = "balls/balls.atlas";
```
Comment (Trulsmatias) [inline core/src/com/gaalf/manager/GameAssetManager.java:20]: Nice! :)

File: core/src/com/gaalf/game/precreatedEntities/balls/Ball.java (hunk 2 above)
Comment (Trulsmatias) [reply core/src/com/gaalf/game/precreatedEntities/balls/Ball.java]: Also: maybe better to just take in the string to create the sprite as a parameter?

**Review by Trulsmatias — COMMENTED** [review]
Looks good :) Look at my comments for further details

File: core/src/com/gaalf/manager/SettingsManager.java (hunk 6 above)
Comment (Jonas-C) [reply core/src/com/gaalf/manager/SettingsManager.java:34]: Default value

File: core/src/com/gaalf/game/precreatedEntities/balls/BallFactory.java (hunk 1 above)
Comment (Jonas-C) [reply core/src/com/gaalf/game/precreatedEntities/balls/BallFactory.java]: The texture atlas file has these values. By naming them with capital letters we can show the texture atlas values directly in the application as opposed to giving each and every one custom properties.

File: core/src/com/gaalf/game/precreatedEntities/balls/Ball.java (hunk 2 above)
Comment (Jonas-C) [reply core/src/com/gaalf/game/precreatedEntities/balls/Ball.java]: Possibly change TextureComponent to SpriteComponent?

File: core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java (hunk 3 above)
Comment (Jonas-C) [reply core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java:30]: Could you elaborate?

File: core/src/com/gaalf/presenter/MainMenuPresenter.java (hunk 7 above)
Comment (Jonas-C) [reply core/src/com/gaalf/presenter/MainMenuPresenter.java:27]: I can change to Squareball, I just thought this looked better

File: core/src/com/gaalf/game/precreatedEntities/balls/BallFactory.java (hunk 1 above)
Comment (Trulsmatias) [reply core/src/com/gaalf/game/precreatedEntities/balls/BallFactory.java]: I see. Make sense. But isn't golf ball for instance written in two words?

File: core/src/com/gaalf/game/precreatedEntities/balls/Ball.java (hunk 2 above)
Comment (Trulsmatias) [reply core/src/com/gaalf/game/precreatedEntities/balls/Ball.java]: Do you mean as in name of the function, or the actual components we are adding?

File: core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java (hunk 3 above)
Comment (Trulsmatias) [reply core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java:30]: To me it looked like all the values of the balls where gathered in on atlas file. Maybe we should gather all these values in one place as well, so we dont need to hard code them in every time

File: core/src/com/gaalf/presenter/MainMenuPresenter.java (hunk 7 above)
Comment (Trulsmatias) [reply core/src/com/gaalf/presenter/MainMenuPresenter.java:27]: To me it looks as if the player is square or something haha

File: core/src/com/gaalf/game/precreatedEntities/balls/BallFactory.java (hunk 1 above)
Comment (Jonas-C) [reply core/src/com/gaalf/game/precreatedEntities/balls/BallFactory.java]: Come on, give a guy a break!!

File: core/src/com/gaalf/game/precreatedEntities/balls/Ball.java (hunk 2 above)
Comment (Jonas-C) [reply core/src/com/gaalf/game/precreatedEntities/balls/Ball.java]: Well both, we'd have a SpriteComponent and an addSpriteComponent

File: core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java (hunk 3 above)
Comment (Jonas-C) [reply core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java:30]: The atlas file only covers details about the placement of a texture in an atlas file, not box2d values

File: core/src/com/gaalf/presenter/MainMenuPresenter.java (hunk 7 above)
Comment (Jonas-C) [reply core/src/com/gaalf/presenter/MainMenuPresenter.java:27]: I can change it to square ball if you wanna

File: core/src/com/gaalf/game/precreatedEntities/balls/Ball.java (hunk 2 above)
Comment (Trulsmatias) [reply core/src/com/gaalf/game/precreatedEntities/balls/Ball.java]: I see now that the TextureComponent only has a sprite as a value, so I think it is a very good idea to rename yes :)

File: core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java (hunk 3 above)
Comment (Trulsmatias) [reply core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java:30]: ah, ok. I don't have too much experience with atlas files. But I still think it is a good idea to gather all the samey values in one place. But maybe in an other pull request?

File: core/src/com/gaalf/presenter/MainMenuPresenter.java (hunk 7 above)
Comment (Trulsmatias) [reply core/src/com/gaalf/presenter/MainMenuPresenter.java:27]: yes please :)

File: core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java (hunk 3 above)
Comment (Jonas-C) [reply core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java:30]: Yeah, that's a refactor waiting to happen.

File: core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java [hunk 9]
```diff
@@ -19,7 +19,7 @@
         TransformComponent transformComponent = addTransformComponent(tiledMap, 1.5f);
         TextureAtlas textureAtlas = assetManager.manager.get(assetManager.ballSpriteAtlas);
         Sprite sprite = textureAtlas.createSprite("Basketball");
```
Comment (Trulsmatias) [inline core/src/com/gaalf/game/precreatedEntities/balls/BasketBall.java:21]: Maybe move this line inside the addSpriteComponent() function?

**Review by Trulsmatias — COMMENTED** [review]
Added one more comment

**Review by Trulsmatias — APPROVED** [review]
Looks good now. Good job! :)

**Review by Brogzybro — APPROVED** [review]
👍

**Pull request merged**